code/game/turfs/change_turf.dm

```diff
@@ -132,6 +132,16 @@ GLOBAL_LIST_INIT(blacklisted_automated_baseturfs, typecacheof(list(
 	QUEUE_SMOOTH_NEIGHBORS(src)
 	QUEUE_SMOOTH(src)
 
+	#if MIN_COMPILER_VERSION >= 514
+	#warn Please replace the loop below this warning with an `as anything` loop.
+	#endif
+
+	for (var/_content_of_turf in contents)
```

Comment:
is there really no way to avoid this loop? change turf can be called a lot, feels like this moves the problem again.

I think you're better off keeping a special list of elements from the old turf and passing to the new turf

Reply:
basically the same way the component handler works, send a signal pre move that fills out a list of things that care about the change and then have them signalled post move.

Reply:
Yeah you should really do that instead

Reply:
I tried this at first, but unless I'm thinking about something wrong this problem comes up:

![image](https://user-images.githubusercontent.com/35135081/113690285-16f3f880-9680-11eb-866a-2f7aee3da552.png)

Reply:
Basically, elements can't really preserve the source through turf changes with the current signal, unless I'm missing something. CC @ninjanomnom

Reply:
We could avoid the loop by hooking to change turf then running some code with whatever the equivalent to `spawn(0)` is so that it defers to the end, but I really want to know if there's a better solution.

Reply:
Just realized the solution oranges made, I suppose that's possible but it sounds a bit funky. Waiting to see if ninja has a better suggestion.

Reply:
This is butting up against what features elements discarded compared to components to be lighter on resources. Elements don't handle managing signals of things other than their target well, a signal on a turf will only let the element know what turf is changing but not what wanted to know as the element doesn't keep track of that.

The only alternative I see is to make this behavior a component or keeping a target list on the element.

Reply:
I think my approach is the correct one.

Reply:
Your approach doesn't solve the problem. At the point in time that ChangeTurf is happening, there is nothing connecting the element with what it has been applied to. Here's the references you have:

```
element <--> turf
listener <--> turf
```

With a signal involving the turf the element needs some other way of getting a reference to the listener. Adding some way in the signal to specify things to transfer doesn't help in getting a reference to the listener.